File: design_patterns_udemy/01_SOLID/SRP.cpp
/*
 * From the Udemy course "Design Patterns in Modern C++"
 * https://www.udemy.com/course/patterns-cplusplus/
 * by Dmitri Nesteruk.
 * 
 * Single Responsibility Principle
 * 
 */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <boost/lexical_cast.hpp>
using namespace std;

struct Journal
{
  string title;
  vector<string> entries;

  // explicit: the constructor must have this form
  // Journal J("my tytle"); // OK
  // Journal J = "my title"; // WRONG
  explicit Journal(const string& title)
    : title{title}
  {
  }

  void add(const string& entry);

  // persistence is a separate concern
  void save(const string& filename);
};

void Journal::add(const string& entry)
{
  // static: 
  // - class members (variables): one exists for the class, all objects share it
  // - class methods (functions): can be used even though no object exists: class::method()
  // - outside a class: it means function/variable are bound only to the scope of the source file,
  //      it exists only there;
  //      advantage: we can repeat names in other files
  // boost::lexical_cast: casting with boost
  static int count = 1; // static class member: called only once, one for the class / all objects share the same
  entries.push_back(boost::lexical_cast<string>(count++) + ": " + entry);
}

// `Journal` focuses only on adding and storing entries in memory
// If we assign to it the responsibility of saving to disk, we're extending
// its responsibilities.
// If we decide to change the saving to be in a data base instead of in a TXT,
// we might want to do it for all similar classes,
// so having everything in a centralized place is very helpful:
// we don't need to change the separate classes, but just the `PersistenceManager`!
void Journal::save(const string& filename)
{
  // ofs: output filestream: we open a file and save lines to it
  ofstream ofs(filename);
  for (auto& s : entries)
    ofs << s << endl;
}

// Principle of Separation of Concerns: each class is concerned with one issue
// See comments in `Journal::save()`
struct PersistenceManager
{
  // static method: can be called even without PM is not instantiated
  // example: PersistenceManager::save(journal, "diary.txt")
  static void save(const Journal& j, const string& filename)
  {
    ofstream ofs(filename);
    for (auto& s : j.entries)
      ofs << s << endl;
  }

  // Other save() functions of similar classes would go here, too

};

int main()
{
  Journal journal{"Dear Diary"};
  journal.add("I ate a bug");
  journal.add("I cried today");

  // Instead of using class-specific method
  // we use PM below
  //journal.save("diary.txt");

  // Since save is static,
  // we can use it without instantiating PM!
  //PersistenceManager pm;
  //pm.save(journal, "diary.txt");
  PersistenceManager::save(journal, "diary.txt");

  return 0;
}

File: design_patterns_udemy/01_SOLID/srp.py
class Journal:
    def __init__(self):
        self.entries = []
        self.count = 0

    def add_entry(self, text):
        self.entries.append(f"{self.count}: {text}")
        self.count += 1

    def remove_entry(self, pos):
        del self.entries[pos]

    def __str__(self):
        return "\n".join(self.entries)

    # Break SRP = Single Responsibility Principle
    # Saving is a responsibility that should be handled
    # by a class that persists the complete family
    # of journal classes!
    # Giving too many responsibilities to single classes
    # creates the anti-pattern of a God Object:
    # a gigant class which does everything,
    # which is very difficult to maintain!
    def save(self, filename):
        file = open(filename, "w")
        file.write(str(self))
        file.close()

    def load(self, filename):
        pass

    def load_from_web(self, uri):
        pass

# Persisting of objects should be handled by a spacific class
# which takes care of all faminily of journal classes.
# That way, the saving processes are all localized
# in the same spot -> easier to maintain!
class PersistenceManager:
    # Static method: it cannot access/modify either the instance or the class
    # but it can be used from both.
    # They signal that the function is independent from the class/object,
    # i.e., some kind of utility procedure; that improves mantainability.
    @staticmethod
    def save_to_file(journal, filename):
        file = open(filename, "w")
        file.write(str(journal))
        file.close()

if __name__ == "__main__":

    j = Journal()
    j.add_entry("I cried today.")
    j.add_entry("I ate a bug.")
    print(f"Journal entries:\n{j}\n")

    p = PersistenceManager()
    file = r'./journal.txt'
    p.save_to_file(j, file)

    # verify!
    with open(file) as fh:
        print(fh.read())
